### PR TITLE
docs(dev guide): lifecycle-hooks - updated dart/ts code and new dart prose

### PR DESCRIPTION
**Pull request by chalin**
**TS** `guide/lifecycle-hooks` updates:
- example code clean-up; **e2e tests pass**
- `.jade` file:
  - mainly copyedits, but also
  - prep for extension by Dart .jade file

**Dart** guide/lifecycle-hooks:
- Significant update of example code (so it matches TS)
- Dart `.jade` extends TS `.jade` file with minor overrides

### CODE REVIEW

**Comment by chalin** [comment]
@thos: could you review the two dart commits?
@filipesilva, or someone else from the team: could you review the two TS commits? The edits are _relatively_ minor.

File: public/docs/dart/latest/_util-fns.jade [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
+mixin liveExampleLink(linkText, exampleUrlPartName)
+  a(href='https://angular-examples.github.io/#{exampleUrlPartName}')= linkText
+
+mixin liveExampleLink2(linkText, exampleUrlPartName)
+  - var liveExampleSourceLinkText = attributes.srcLinkText || 'view source'
+  span.
+    #[+liveExampleLink(linkText, exampleUrlPartName)]
+    (#[a(href='https://github.com/angular-examples/#{exampleUrlPartName}') #{liveExampleSourceLinkText}])
+    
+//- Depricated
```
Comment (Foxandxss) [inline public/docs/dart/latest/_util-fns.jade]: Deprecated

**Comment by Foxandxss** [comment]
I like the TS changes Patrice. Lot of good stuff in here. One question: What are those new blocks you are defining there?

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+class ChildComponent {
+  String hero = 'Magneta';
+}
+
+//////////////////////
+@Component(
+    selector: 'after-content',
+// #docregion template
+    template: '''
+    <div>-- projected content begins --</div>
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Indent the template HTML by 2 more spaces.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      <div *ngIf="show">
+        <after-content>
+          <my-child></my-child>
+        </after-content>
+      </div>
+
+    <h4>-- AfterContent Logs --</h4>
+    <p><button (click)="reset()">Reset</button></p>
+    <div *ngFor="let msg of logs">{{msg}}</div>
+  </div>
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Indentation is off here + indent template by 2 more spaces.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  }
+// #enddocregion do-something
+
+  void _logIt(String method) {
+    var child = contentChild;
+    var message = "${method}: ${child != null ? child.hero:'no'} child content";
+    _logger.log(message);
+  }
+// #docregion hooks
+  // ...
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: What's this?

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+// #docregion do-something
+  // This surrogate for real business logic sets the `comment`
+  _doSomething() {
+    comment = contentChild.hero.length > 10 ? "That's a long name" : '';
+  }
+// #enddocregion do-something
+
+  void _logIt(String method) {
+    var child = contentChild;
+    var message = "${method}: ${child != null ? child.hero:'no'} child content";
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Prefer using nullaware operators for these:
  `child?.hero ?? 'no'`

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      _prevHero = contentChild?.hero;
+      _logIt('AfterContentChecked');
+      _doSomething();
+    }
+  }
+// #enddocregion hooks
+// #docregion do-something
+  // This surrogate for real business logic sets the `comment`
+  _doSomething() {
+    comment = contentChild.hero.length > 10 ? "That's a long name" : '';
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Prefer single quotes for Strings in Dart.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      _logIt('AfterContentChecked (no change)');
+    } else {
+      _prevHero = contentChild?.hero;
+      _logIt('AfterContentChecked');
+      _doSomething();
+    }
+  }
+// #enddocregion hooks
+// #docregion do-something
+  // This surrogate for real business logic sets the `comment`
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Doc comments start with triple slash: `///`

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    <p><button (click)="reset()">Reset</button></p>
+    <div *ngFor="let msg of logs">{{msg}}</div>
+  </div>
+    ''',
+// #enddocregion parent-template
+    styles: const ['.parent {background: burlywood}'],
+    providers: const [LoggerService],
+    directives: const [AfterContentComponent, ChildComponent])
+class AfterContentParentComponent {
+  LoggerService _logger;
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Make injected deps `final`

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  List<String> logs;
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Nit: This could be a getter on the _logger.logs instead.

**Comment by chalin** [comment]
@Foxandxss : thanks for reviewing. The Jade `block`s define text (usually a few paragraphs) that are potential points of variation between the doc versions (TS, Dart, JS); a bit like class methods which can be overridden in a subclass. In this case, for the Dart docs, we `extend` the TS doc and override the blocks with Dart specific text.

File: public/docs/dart/latest/_util-fns.jade (hunk 1 above)
Comment (chalin) [inline public/docs/dart/latest/_util-fns.jade]: Fixed

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 2 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 3 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: fixed

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 4 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: The `...`? It is used in the prose to indicate that there is more in that class but it is not being shown. That convention is used on the TS side.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 5 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Ah yes. Cleaner. Done.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 6 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: I generally do, except when it contains a single single-quote like this.
I find `"That's"` easier to read than `'That\'s'`.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 7 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 6 above)
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: I'm blind, didn't see the quote.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 8 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: fixed (everywhere in example)

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart (hunk 9 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_content_component.dart]: Yes, that is cleaner. Fixed (everywhere in example).

**Comment by filipesilva** [comment]
Sorry for the delay @chalin. I'm with @foxandxss, the TS bits look good to me.

**Comment by chalin** [comment]
@filipesilva : np, thanks.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart [hunk 10]
```diff
@@ -1,77 +1,117 @@
...
+  directives: const [ChildViewComponent])
+// #docregion hooks
+class AfterViewComponent implements AfterViewChecked, AfterViewInit {
+  var _prevHero = '';
+  
+  // Query for a VIEW child of type `ChildViewComponent`
+  @ViewChild(ChildViewComponent) ChildViewComponent viewChild;
+  
+// #enddocregion hooks
   LoggerService _logger;
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: Newline before constructor.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart [hunk 11]
```diff
@@ -1,77 +1,117 @@
...
-  List<String> hookLog;
-  String hero = 'Magneta';
-  bool showChild = true;
+  AfterViewComponent(this._logger) {
+    _logIt('AfterView constructor');
+  }
+
+// #docregion hooks
+  ngAfterViewInit() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: Please add return types for every function you define.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart [hunk 12]
```diff
@@ -1,77 +1,117 @@
...
+    _logIt('AfterViewInit');
+    _doSomething();
+  }
 
-  // Query for a CONTENT child of type `ChildComponent`
-  // No such CONTENT child exists!
-  // This component holds a view but no content of that type.
-  @ContentChild(ChildComponent)
-  ChildComponent contentChild;
+  ngAfterViewChecked() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart [hunk 13]
```diff
@@ -1,77 +1,117 @@
...
 
-  // Query for a VIEW child of type `ChildComponent`
-  @ViewChild(ChildComponent)
-  ChildComponent viewChild;
+  String comment = '';
 
-  String _prevHero;
+// #docregion do-something
+  // This surrogate for real business logic sets the `comment`
+  _doSomething() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart [hunk 14]
```diff
@@ -68,10 +66,14 @@ class CounterParentComponent {
     reset();
   }
 
-  updateCounter() => value += 1;
+  updateCounter() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart [hunk 15]
```diff
@@ -68,10 +66,14 @@ class CounterParentComponent {
...
+    value += 1;
+    _logger.tick();
+  }
 
   reset() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart]: Ditto, return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+  bool changeDetected = false;
+  List<String> changeLog = [];
+
+  String oldHeroName = '';
+  String oldPower = '';
+  int oldLogLength = 0;
+  int noChangeCount = 0;
+
+  // #docregion ng-do-check
+  ngDoCheck() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        changeLog[changeLog.length - 1] = noChangeMsg;
+      }
+    }
+
+    changeDetected = false;
+  }
+  // #enddocregion ng-do-check
+
+  // Copied from OnChangesComponent
+  ngOnChanges(Map<String, SimpleChange> changes) {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    changes.forEach((String propName, SimpleChange change) {
+      String cur = JSON.encode(change.currentValue);
+      String prev =
+          change.isFirstChange() ? "{}" : JSON.encode(change.previousValue);
+      changeLog.add('$propName: currentValue = $cur, previousValue = $prev');
+    });
+  }
+
+  reset() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+    changeDetected = false;
+  }
+  // #enddocregion ng-do-check
+
+  // Copied from OnChangesComponent
+  ngOnChanges(Map<String, SimpleChange> changes) {
+    changes.forEach((String propName, SimpleChange change) {
+      String cur = JSON.encode(change.currentValue);
+      String prev =
+          change.isFirstChange() ? "{}" : JSON.encode(change.previousValue);
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Is this the `dartfmt` formatting? If not could you run it on this file?

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+  templateUrl: 'on_changes_parent_component.html',
+  styles: const ['.parent {background: Lavender}'],
+  directives: const [DoCheckComponent])
+class DoCheckParentComponent {
+  Hero hero;
+  String power;
+  String title = 'DoCheck';
+  @ViewChild(DoCheckComponent) DoCheckComponent childView;
+
+  DoCheckParentComponent() { reset(); }
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: `dartfmt` this

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+  reset() {
```
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Return type

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart (hunk 11 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: It has an implicit return type of `dynamic`, thus matching the signature of the interface method it is overriding. I would suggest leaving it like this.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart (hunk 12 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: It has an implicit return type of `dynamic`, thus matching the signature of the interface method it is overriding. I would suggest leaving it like this.

File: public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart (hunk 13 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/after_view_component.dart]: fixed (set to `void`)

File: public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart (hunk 14 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart (hunk 15 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 16 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: It has an implicit return type of `dynamic`, thus matching the signature of the interface method it is overriding. I would suggest leaving it like this.

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 17 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: It has an implicit return type of `dynamic`, thus matching the signature of the interface method it is overriding. I would suggest leaving it like this.

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 18 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 19 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 20 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: done

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 21 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: done

**Comment by chalin** [comment]
@thso: all done with the suggested changes --- except for adding return types to methods whose supertype signatures leave the return type implicitly `dynamic`.

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 16 above)
Comment (thso) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: When the return type is dynamic prefer adding it explicitly.

**Comment by thso** [comment]
LGTM

File: public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart (hunk 16 above)
Comment (chalin) [inline public/docs/_examples/lifecycle-hooks/dart/lib/do_check_component.dart]: Ok

**Pull request merged**